Add message parse mode (code vs text) design doc (#474)
unicode-org/message-format-wg: exploration/text-vs-code.md

# Message Parse Mode

Status: **Proposed**

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@eemeli</dd>
		<dd>@aphillips</dd><!-- Seville and other inserted edits -->
		<dt>First proposed</dt>
		<dd>2023-09-13</dd>
		<dt>Pull Request</dt>
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/474">#474</a></dd>
	</dl>
</details>

## Objective

> Decide whether text patterns or code statements should be enclosed in MF2.

Author: The "how" part would be an extension of what the design doc is currently doing. Is that intentional?

> Decide whether text patterns or code statements should be enclosed in MF2.

Reviewer: You're right, but this isn't about "whether text patterns or code statements should be enclosed" but rather about which should be enclosed and when. And the design decision is really about the general syntax (text-vs-code and trimmed-vs-untrimmed-vs-quoted)

and [Android String Resources](https://developer.android.com/guide/topics/resources/string-resource.html#Plurals) in XML.
These formats rely on the resource format providing clear delineation of the beginning and end of a pattern.

Based on available data,
no more than 0.3% of all messages and no more than 0.1% of messages with variants
contain leading or trailing whitespace.
No more than one third of this whitespace is localizable,
and most commonly it's due to improper segmentation or other internationalization bugs.

## Use-Cases

Most messages in any localization system do not contain any expressions, statements or variants.
These should be expressible as easily as possible.

Many messages include expressions that are to be interpolated during formatting.
For example, a greeting like "Hello, user!" may be formatted in many locales with the `user`
being directly set by an input variable.

Sometimes, interpolated values need explicit formatting within a message.
For example, formatting a message like "You have eaten 3.2 apples"
may require the input numerical value
to be formatted with an explicit `minimumFractionDigits` option.

> Some messages require multiple variants.
> This is often related to plural cases, such as "You have 3 new messages",
> where the value `3` is an input and the "messages" needs to correspond with its plural category.

Reviewer: Responding to @echeran's comment in yesterday's call, we should be a bit fastidious about saying why you can't just wrap `{$numMessages} new messages` with a plural ("concatentation folly"). I don't think we need to fully explain it here, but we should use terminology consistent with our "no concatenation" ethos.

```suggestion
Some messages need to choose between multiple patterns (called "variants").
For example, this is often related to the handling of numeric values,
in which the pattern used for formatting depends on one of the data values
according to its plural category (see [CLDR Plural Rules](https://cldr.unicode.org/index/cldr-spec/plural-rules)
for more information). 
So, in English, the formatter might need to choose between formatting
`You have 1 kilometer to go` and `You have 2 kilometers to go` 
```

Multiple levels of complex nesting should be avoided,
along with any constructs that require an excessive
level of precision on the part of non-technical users.

As MessageFormat 2 will be at best a secondary language to all its users,
it should conform to user expectations and require as little learning as possible.

The syntax should avoid footguns,
in particular as it's passed through various tools during formatting.

ASCII-compatible syntax. While support for non-ASCII characters for variable names,
values, literals, options, and the like are important, the syntax itself should
be restricted to ASCII characters. This allows the message to be parsed
visually by humans even when embedded in a syntax that requires escaping.

Whitespace is forgiving. We _require_ the minimum amount of whitespace and allow
users to format or change unimportant whitespace as much as they want.
This avoids the need for translators or tools to be super pedantic about
formatting.

## Constraints

Limiting the range of characters that need to be escaped in plain text is important.
Following past precedent,
this design doc will only consider encapsulation styles which

> start with `{` and end with `}`.

Reviewer: This isn't actually true. We should remove it, especially since it's not germane to text-vs-code or the different PEWS handling options.

Author: ```suggestion
```

Encapsulate with `{…}` or otherwise distinguishing statements from
the primarily unquoted translatable message contents.

For messages with multiple variants,
separate the variants using `when` statements.

Trim whitespace between and around statements such as `input` and `when`,
but do not otherwise trim any leading or trailing whitespace from a message.
This allows for whitespace such as spaces and newlines to be used outside patterns
to make a message more readable.

Allow for a pattern to be `{{…}}` quoted
such that it preserves its leading and/or trailing whitespace
even when preceded or followed by statements.

## Alternatives Considered

### Start in code, encapsulate text

This approach treats messages as something like a resource format for pattern values.
Keywords are declared directly at the top level of a message,
and patterns are always surrounded by `{{…}}` or some other delimiters.

Whitespace in patterns is never trimmed.

The `{{…}}` are required for all messages,
including ones that only consist of text.
Delimiters of the resource format are required in addition to this,
so messages may appear wrapped as e.g. `"{{…}}"`.

> This option is not chosen due to adding an excessive
> quoting burden on all messages.

Reviewer: I think we should not include these "This option is not chosen due..." paragraphs. I think it is okay to call out objective or subjective reasons for why we might not choose a given alternative, e.g.

```suggestion
- This option makes plain text strings invalid as messages.
- This option requires additional quoting for simple messages.
```

Our choice section should deal with the logic of why a given option was chosen.

Author: Could you clarify what you mean by "choice section"? I'm not sure that I understand what that is.

### Start in text, encapsulate code, re-encapsulate text within code

As in the proposed design, simple patterns are unquoted.
Patterns in messages with statements, however,
are required to always be surrounded by `{{…}}` or some other delimiters.

This effectively means that some syntax will "enable" code mode for a message,
and that patterns in such a message need delimiters.

This option is not chosen due to adding an excessive
quoting burden on all multi-variant messages,
as well as introducing an unnecessary additional conceptual layer to the syntax.

> ### Start in text, encapsulate code, trim minimally

Reviewer: We need to include the missing option, specifically:

**Start in text, encapsulate code, no trimming**

This would be effectively "2a without quotes" in which all **_pattern_** whitespace is meaningful. Perhaps no one would go for this option, but we should have it for completeness.

Author: Added.

This is the same as the proposed design,
but with a different trimming rule:

- Trim all spaces before and between declarations.
- For single-variant messages, trim one newline after the last declaration.
- For multivariant messages,
  trim one space after a `when` statement and
  one newline followed by any spaces before a subsequent `when` statement.

This option is not chosen due to the quoting being too magical.
Even though this allows for all patterns with whitespace to not need quotes,
the cost in complexity is too great.

### Start in text, encapsulate code, trim maximally

This is the same as the proposed design,
but with a different trimming rule:

- Trim all leading and trailing whitespace for each pattern.

Expressing the trimming on patterns rather than statements
means that leading and trailing spaces are also trimmed from simple messages.
This option is not chosen due to this being somewhat surprising,
especially when messages are embedded in host formats that have predefined means

> of escaping and/or trimming leading and trailing spaces from a value.

Reviewer: Trimming simple patterns like this is a bridge too far for me. It should be a separate decision for the "trim XXX" options whether they are trimmed. I can make a plausible argument for why simple patterns should behave differently when trimmed than variant patterns do.

Based on available data,
no more than 0.3% of all messages and no more than 0.1% of messages with variants
contain leading or trailing whitespace.
No more than one third of this whitespace is localizable,
and most commonly it's due to improper segmentation or other internationalization bugs.

## Use-Cases

Most messages in any localization system do not contain any expressions, statements or variants.
These should be expressible as easily as possible.

Many messages include expressions that are meant to be replaced during formatting.
For example, a greeting like "Hello, {$username}!" would be formatted with the variable
`$username` being replaced by an input variable.

In some rare cases, replacement variables might be added (or removed) in one particular
locale versus messages in other locales.

Sometimes, the replacement variables need to be formatted.
For example, formatting a message like `You have {$distance} kilometers to go`
requires that the numeric value `{$distance}` be formatted as a number according to the locale: `You have 1,234 kilometers to go`.

Formatting of replacement variables might also require tailoring.

> For example, if the user wants to show fractions of a kilometer in the above example

Reviewer 2: Nit: This use of "user" has a different meaning than the use in «_a greeting like "Hello, {$username}!"_» three paragraphs earlier. We haven't been great about vocabulary here, but I think it would make sense to explicitly reference the [Web Platform Priority of Constituencies](https://www.w3.org/TR/design-principles/#priority-of-constituencies) ("user" vs. "author" vs. "implementor" vs. "specification writer" vs. "theoretical purity") or define and use analogous terms (e.g., "developer" and "translator" rather than "author").

Author: I've refactored the text to refer to "authors", "developers" and/or "translators" rather than "users".

I think bringing in the priority of constituencies is beyond the scope of this design doc, but it could be valuable to introduce separately.

formatting.

## Constraints

Limiting the range of characters that need to be escaped in plain text is important.

The current syntax includes some plain-ascii keywords:
`input`, `local`, `match`, and `when`.

The current syntax and active proposals include some sigil + name combinations,
such as `:number`, `$var`, `|literal|`, `+bold`, and `@attr`.

The current syntax supports unquoted literal values as operands.

Messages themselves are "simple strings" and must be considered to be a single
line of text. In many containing formats, newlines will be represented as the local
equivalent of `\n`.

## Proposed Design

### Start in text, encapsulate code, trim around statements

Allow for message patterns to not be quoted.

Encapsulate with `{…}` or otherwise distinguishing statements from
the primarily unquoted translatable message contents.

For messages with multiple variants,
separate the variants using `when` statements.

Trim whitespace between and around statements such as `input` and `when`,
but do not otherwise trim any leading or trailing whitespace from a message.
This allows for whitespace such as spaces and newlines to be used outside patterns

> to make a message more readable.

Reviewer 2: I'm concerned about this resulting in surprising misbehavior... consider a developer making the following change, in which removing all inputs from a message that still starts and ends with a line feed would not be expected to affect whitespace:
`````diff
logOutMessage = ```
-{%input username}
-
-Log out {$username}?
+Log out?
```;
`````

Author: Ah, fair point. The issue here is that the above edit would add a new line to the message's start, yes? One way to avoid this would be to require a message with statements to not have leading whitespace. That might be a reasonable restriction.

Reviewer: I agree with @gibson042's illustration being informative. I would have guessed (if I didn't know anything about MF2) that the "before" of `logOutMessage` included a newline. With some of the options here, the newline after `{%input $username}` is consumed, but not the blank link after that. In other options both newlines are consumed unless the user specifically quoted the pattern or whitespace.

Also, note that many formats would encode the example as:

```
// using our current syntax but without quoting the pattern
var logOutMessage = "{{{#input $username}\n\nLog out {$username}?}}
```
